### PR TITLE
Extract CTF Input Data to Separate Structure

### PR DESCRIPTION
**Pull request by bska**
This commit creates a new helper structure,
```
Opm::Connection::CTFProperties
```
that holds all of the individual items which go into calculating the final connection transmissibility factor (CTF).  These include the `Kh` product, the pressure equivalent radius (`r0`), the well-bore radius (`rw`), and the skin factor.  We reimplement the `Connection` constructor in terms of this new helper structure and this changes the API of the constructor.  As an additional helper for the implementation, and with a view towards simulation restart support, we also store the value of the Peaceman formula denominator in CTFProperties as
```C++
double CTFProperties::peaceman_denom
```
This, in turn, simplifies the implementation of `CSKIN`.  It also fixes a latent problem in the current `CSKIN` implementation which would ignore any explicitly input CTF values in favour of the pressure equivalent radius.  By updating the explicit denominator with a new skin factor instead we do not need to recompute the logarithmic term involving `r0` as part of `CSKIN` handling.

### CODE REVIEW

**Comment by bska** [comment]
Note: Since this PR changes the `Connection` constructor it will shortly have a downstream counterpart to update all such constructor calls.

**Comment by bska** [comment]
jenkins build this opm-simulators=4989 please

File: src/opm/input/eclipse/Schedule/Well/Connection.cpp [hunk 1]
```diff
@@ -509,41 +597,35 @@ void Connection::setInjMult(const InjMult& inj_mult) {
     m_injmult = inj_mult;
 }
 
-
 void Connection::setFilterCake(const FilterCake& filter_cake) {
```
Comment (akva2) [inline src/opm/input/eclipse/Schedule/Well/Connection.cpp:600]: any reason you stopped moving open braces to a new line here?

Comment (bska) [reply src/opm/input/eclipse/Schedule/Well/Connection.cpp]: > any reason you stopped moving open braces to a new line here?

No, not really.  I've pushed an update to make that change (function's opening brace on the next line) throughout the file.  Now I "just" need to figure out why the [parallel regression test](https://ci.opm-project.org/job/opm-common-PR-builder/lastCompletedBuild/testReport/(root)/mpi/compareParallelSim_flow_WINJDAM_MSW/) fails.

**Comment by bska** [comment]
jenkins build this opm-simulators=4989 please

**Comment by bska** [comment]
jenkins build this opm-simulators=4989 please

**Comment by bska** [comment]
I have looked into the [failing regression tests](https://ci.opm-project.org/job/opm-common-PR-builder/6439/testReport/) and apart from `WINJDAM_STDW`, all is good or explained fixing a latent problem in the `CSKIN` implementation.

Case | Status | Comments
------- | :--------: | --------
[2D_OILWATER_POLYMER](https://github.com/OPM/opm-tests/blob/bf2f98f43f11790d0666448b5079fc6e7e23b21e/polymer_oilwater/2D_OILWATER_POLYMER.DATA) | :white_check_mark: |
[2D_POLYMER_INJECTIVITY](https://github.com/OPM/opm-tests/blob/bf2f98f43f11790d0666448b5079fc6e7e23b21e/polymer_injectivity/2D_POLYMER_INJECTIVITY.DATA) | :white_check_mark: | Somewhat visible changes to `INJE01`'s BHP
[2D_THREEPHASE_POLY_HETER](https://github.com/OPM/opm-tests/blob/bf2f98f43f11790d0666448b5079fc6e7e23b21e/polymer_simple2D/2D_THREEPHASE_POLY_HETER.DATA) | :white_check_mark: |
[CSKIN-01](https://github.com/OPM/opm-tests/blob/bf2f98f43f11790d0666448b5079fc6e7e23b21e/cskin/CSKIN-01.DATA) | :white_check_mark: | Affected by "latent problem" fixed in this PR
[CSKIN-02](https://github.com/OPM/opm-tests/blob/bf2f98f43f11790d0666448b5079fc6e7e23b21e/cskin/CSKIN-02.DATA) | :white_check_mark: | Affected by "latent problem" fixed in this PR
[CSKIN-04](https://github.com/OPM/opm-tests/blob/bf2f98f43f11790d0666448b5079fc6e7e23b21e/cskin/CSKIN-04.DATA) | :white_check_mark: | Affected by "latent problem" fixed in this PR
[WINJDAM_STDW](https://github.com/OPM/opm-tests/blob/bf2f98f43f11790d0666448b5079fc6e7e23b21e/winjdam/WINJDAM_STDW.DATA) | :x: | Significant differences, even with `--enable-tuning=true`.  See below.

#### Case Comparison `WINJDAM_STDW` ####

I **believe** that the changes here are caused by using the Peaceman formula denominator [directly](https://github.com/OPM/opm-simulators/blob/f481c5c55215d882d1cce2e2a9be6a94fd5cd0f6/opm/simulators/wells/WellFilterCake.cpp#L131) instead of [calculating the denominator value from geometric factors](https://github.com/OPM/opm-simulators/blob/c6bb607cdc30ec108b445ef0ac7a6f50b2e1fe7e/opm/simulators/wells/WellFilterCake.cpp#L134-L136) **and** from using a [more correct definition](https://github.com/OPM/opm-common/blob/a4aa75e06075fe585e729d8b1e897c29e18199f9/src/opm/input/eclipse/Schedule/Well/WellConnections.cpp#L488-L489) of the connection interval.  The new definition accounts for input values of the connection transmissibility factor and the `Kh` product.  I'd nevertheless like some input from @GitPaean on the injection damage case, just to make sure that I've not overlooked something.

Comparison | Curves
------- | ----------
Field Level | injdam-field_level
Well `INJ-1` | injdam-well-inj1
Well `PROD-1` | injdam-well-prod1
Connection INJ-1:6,7,3 | injdam-conn-inj1-673
Connection INJ-1:6,7,4 | injdam-conn-inj1-674
Connection INJ-1:6,7,5 | injdam-conn-inj1-675
Connection INJ-1:6,7,6 | injdam-conn-inj1-676
Connection INJ-1:6,7,7 | injdam-conn-inj1-677

**Comment by GitPaean** [comment]
Sorry for not seeing this earlier, was out of office. I will check the plotting by Monday.

**Comment by GitPaean** [comment]
> all is good or explained fixing a latent problem in the CSKIN implementation.

Do you mean the CSKIN development actually broke those tests when we merged it?  I would think the cases without CSKIN keyword should not be affected by the CSKIN development, right?

**Comment by bska** [comment]
> > all is good or explained fixing a latent problem in the CSKIN implementation.
> 
> Do you mean the `CSKIN` development actually broke those tests when we merged it? I would think the cases without `CSKIN` keyword should not be affected by the `CSKIN` development, right?

Right, sorry for not being clear.  The `CSKIN-*` regression failures here are explained by fixing the latent problem in the `CSKIN` implementation.  The other failures, apart from `WINJDAM_STDW`, are all very minor and safe for when we update the reference solutions.  The only real failure is `WINJDAM_STDW`.  I **can** restore that case if I switch back to using `D[2]` as the connection length, but that essentially means reintroducing the same kind of latent error that I just removed from the `CSKIN` implementation; namely that the keyword does not honour the case of `COMPDAT` specifying both the CTF and the Kh product.

**Comment by GitPaean** [comment]
jenkins build this opm-simulators=4989 please

**Comment by bska** [comment]
jenkins build this opm-simulators=4989 please

**Comment by GitPaean** [comment]
> I believe that the changes here are caused by using the Peaceman formula denominator [directly](https://github.com/OPM/opm-simulators/blob/f481c5c55215d882d1cce2e2a9be6a94fd5cd0f6/opm/simulators/wells/WellFilterCake.cpp#L131) instead of [calculating the denominator value from geometric factors](https://github.com/OPM/opm-simulators/blob/c6bb607cdc30ec108b445ef0ac7a6f50b2e1fe7e/opm/simulators/wells/WellFilterCake.cpp#L134-L136) and from using a [more correct definition](https://github.com/OPM/opm-common/blob/a4aa75e06075fe585e729d8b1e897c29e18199f9/src/opm/input/eclipse/Schedule/Well/WellConnections.cpp#L488-L489) of the connection interval.

Thanks for the explanation. I think they should probably be fine. I will request the people using these keywords to test.

**Comment by GitPaean** [comment]
side track a little bit, not related to this PR. I checked the results of 2D_THREEPHASE_POLY_HETER a little bit, it looks like somehow, we are deviating from the reference results quite far now. The black line is the reference results. We probably had perfect matching some time ago.
It is possible when we updated the reference results, we gradually changed the reference results without knowing it. 

image

**Comment by bska** [comment]
> I checked the results of `2D_THREEPHASE_POLY_HETER` a little bit, it looks like somehow, we are deviating from the reference results quite far now. [...] We probably had perfect matching some time ago.  It is possible when we updated the reference results, we gradually changed the reference results without knowing it.

Yeah, I think so too.  It's not clear to me when those deviations were introduced.

> The black line is the reference results.

Okay.  And what are the green and red curves, respectively?  Green is master and red is this PR?  Other way around?  Something else?

**Comment by GitPaean** [comment]
>Okay. And what are the green and red curves, respectively? Green is master and red is this PR? Other way around? Something else?

Green line (master) and red lines (this PR) are on top each other. This PR does not change results at all for this case. 

And also, it is a slightly different version (injection concentration 1.5 instead of 1.0) from the one in the opm-tests, because this one I have reference result. I expect the comparison will be similar.

**Comment by GitPaean** [comment]
> Understood. So do you support merging this PR as is, or would you like us to investigate the polymer differences first?

I am waiting some confirmation from the WINJDAM users.

**Comment by bska** [comment]
> > Understood. So do you support merging this PR as is, or would you like us to investigate the polymer differences first?
> 
> I am waiting for some confirmation from the `WINJDAM` users.

Very good.  I'll keep the PR up-to-date with other development until we hear from them.  In the meantime I'll reset the PR back to draft state to prevent accidental merging.

**Comment by GitPaean** [comment]
jenkins build this opm-simulators=4989 please

**Comment by bska** [comment]
jenkins build this opm-simulators=4989 please

**Comment by GitPaean** [comment]
> I am waiting for some confirmation from the WINJDAM users.

Referring to the private communication, the approach introduced in this PR is favored by the users of WINJDAM. Please go ahead to start the merging process of the PR.

**Comment by bska** [comment]
jenkins build this opm-simulators=4989 update_data please

**Comment by bska** [comment]
jenkins build this opm-simulators=4989 opm-tests=1120 please

**Comment by bska** [comment]
jenkins build this opm-simulators=4989 opm-tests=1120 please

**Comment by bska** [comment]
jenkins build this opm-simulators=4989 opm-tests=1120 please

**Comment by bska** [comment]
jenkins build this opm-simulators=4989 opm-tests=1120 please

**Comment by bska** [comment]
> > Referring to the private communication, the approach introduced in this PR is favored by the users of WINJDAM. Please go ahead to start the merging process of the PR.
> 
> Very good. I'll mark this as "ready for review" and run a final build check.

The build checks are complete.  I'll merge this PR series now.

**Comment by bska** [comment]
The new reference solutions have been installed on the CI system.  I'll merge this now to switch to the new CTF implementation.

**Pull request merged**